x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/whitelist/InternalSqlScriptUtils.java

```diff
@@ -355,11 +359,14 @@ public static ZonedDateTime asDateTime(Object dateTime) {
         if (dateTime instanceof ZonedDateTime) {
             return (ZonedDateTime) dateTime;
         }
-        if (dateTime instanceof Number) {
-            return DateUtils.of(((Number) dateTime).longValue());
+        if (false == lenient) {
```

Comment:
Keep the instanceof check clean and move the if on the throw:

```
if (dateTime instanceof ...)
if (dateTime instanceof ...)
if (dateTime instanceof ...)

if (lenient) { return dateTime; } else { throw exception }
```

Reply:
@costin if I take the `instanceof Number` condition out of the `if (lenient)` part, there will be failures for queries that order by or group by a casted numeric function. This is because the `asDateTime` is specifically being used for date/time values whereas, in case of CONVERT or CAST the `instanceof Number` part will be applied for NON-date/time numerics, as well.